File: packages/node/src/Seguranca/SuiteDeAtaques.js
'use strict';

const ResultadoAtaque = require('./ResultadoAtaque.js');
const SkipAtaqueException = require('./SkipAtaqueException.js');

class SuiteDeAtaques {
  constructor() {
    this.ataques = [];
  }

  adicionar(ataque) {
    this.ataques.push(ataque);
    return this;
  }

  /** @returns {ResultadoAtaque[]} */
  rodar(alvo) {
    const resultados = [];
    for (const ataque of this.ataques) {
      try {
        resultados.push(ataque.executar(alvo));
      } catch (e) {
        if (e instanceof SkipAtaqueException) {
          resultados.push(
            new ResultadoAtaque(ataque.nome(), false, 'pulado', 'Pulado: ' + e.message),
          );
        } else {
          resultados.push(
            new ResultadoAtaque(ataque.nome(), true, 'erro', 'Erro ao executar: ' + e.message),
          );
        }
      }
    }
    return resultados;
  }

  /** Roda e imprime um relatório direto no console. @returns {boolean} */
  rodarEImprimir(alvo) {
    const resultados = this.rodar(alvo);
    let vulnerabilidades = 0;

    console.log('='.repeat(70));
    console.log('RELATÓRIO DA SUÍTE DE ATAQUES');
    console.log('='.repeat(70));
    console.log();

    for (const r of resultados) {
      console.log(r.linhaResumo());
      if (r.vulneravel && r.severidade !== 'pulado' && r.severidade !== 'demonstracao') {
        vulnerabilidades++;
      }
    }

    console.log();
    console.log('='.repeat(70));
    if (vulnerabilidades === 0) {
      console.log(`RESUMO: nenhuma vulnerabilidade encontrada em ${resultados.length} ataque(s).`);
    } else {
      console.log(
        `RESUMO: ${vulnerabilidades} vulnerabilidade(s) encontrada(s) de ${resultados.length} ataque(s) rodados!`,
      );
    }
    console.log('='.repeat(70));

    return vulnerabilidades === 0;
  }
}

module.exports = SuiteDeAtaques;

File: packages/rust/src/seguranca/suite_de_ataques.rs
//! Runner da suíte de ataques.

use crate::seguranca::alvo_criptografico::AlvoCriptografico;
use crate::seguranca::ataque::{Ataque, ErroAtaque};
use crate::seguranca::resultado_ataque::{ResultadoAtaque, Severidade};

pub struct SuiteDeAtaques {
    ataques: Vec<Box<dyn Ataque>>,
}

impl SuiteDeAtaques {
    pub fn new() -> Self {
        SuiteDeAtaques {
            ataques: Vec::new(),
        }
    }

    pub fn adicionar(&mut self, ataque: Box<dyn Ataque>) -> &mut Self {
        self.ataques.push(ataque);
        self
    }

    pub fn rodar(&self, alvo: &dyn AlvoCriptografico) -> Vec<ResultadoAtaque> {
        let mut resultados = Vec::new();
        for ataque in &self.ataques {
            let nome = ataque.nome();
            match ataque.executar(alvo) {
                Ok(resultado) => resultados.push(resultado),
                Err(ErroAtaque::Skip(mensagem)) => resultados.push(ResultadoAtaque::new(
                    &nome,
                    false,
                    Severidade::Pulado,
                    &format!("Pulado: {}", mensagem),
                )),
                Err(ErroAtaque::Falha(mensagem)) => resultados.push(ResultadoAtaque::new(
                    &nome,
                    true,
                    Severidade::Erro,
                    &format!("Erro ao executar: {}", mensagem),
                )),
            }
        }
        resultados
    }

    /// Roda e imprime um relatório direto no console.
    pub fn rodar_e_imprimir(&self, alvo: &dyn AlvoCriptografico) -> bool {
        let resultados = self.rodar(alvo);
        let mut vulnerabilidades = 0usize;

        let barra = "=".repeat(70);
        println!("{}", barra);
        println!("RELATÓRIO DA SUÍTE DE ATAQUES");
        println!("{}", barra);
        println!();

        for r in &resultados {
            println!("{}", r.linha_resumo());
            if r.vulneravel
                && r.severidade != Severidade::Pulado
                && r.severidade != Severidade::Demonstracao
            {
                vulnerabilidades += 1;
            }
        }

        println!();
        println!("{}", barra);
        if vulnerabilidades == 0 {
            println!(
                "RESUMO: nenhuma vulnerabilidade encontrada em {} ataque(s).",
                resultados.len()
            );
        } else {
            println!(
                "RESUMO: {} vulnerabilidade(s) encontrada(s) de {} ataque(s) rodados!",
                vulnerabilidades,
                resultados.len()
            );
        }
        println!("{}", barra);

        vulnerabilidades == 0
    }
}

impl Default for SuiteDeAtaques {
    fn default() -> Self {
        Self::new()
    }
}
